File: src/python/ui/components/audio.py
#!/usr/bin/env python
#
# audio.py
#
# Audio option panel for SDRLibEConsole
# 
# Copyright (C) 2019 by G3UKB Bob Cowdery
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#    
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#    
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#    
#  The author can be reached by email at:   
#     bob@bobcowdery.plus.com
#

# Import all
from main.imports import *

#==============================================================================================
# Audio provides a selection of available options per receiver and transmitter
#==============================================================================================

#=====================================================
# Audio main class
#===================================================== 
class Audio(QWidget):
    
    #-------------------------------------------------
    # Constructor
    def __init__(self, id):
        """
        Constructor
        
        Arguments:
            id  -- the radio id
            
        """
        super(Audio, self).__init__()
        
        # radio id
        self.__id = id
        
        # Get the connector instance
        self.__con = getInstance('conn_inst')
        
        # Create the UI
        # Set the back colour
        palette = QPalette()
        palette.setColor(QPalette.Background, QColor(41,41,41,255))
        self.setPalette(palette)
        self.setWindowFlags(
            Qt.CustomizeWindowHint |
            Qt.FramelessWindowHint |
            Qt.WindowStaysOnTopHint
        )
        self.setWindowOpacity(0.97)
        # Set the layout
        grid = QGridLayout()
        self.setLayout(grid)
        
        # Get radio model audio section
        self.__radio_1_audio_model = Model.get_radio_model()[1]['AUDIO']
        self.__radio_2_audio_model = Model.get_radio_model()[2]['AUDIO']
        self.__radio_3_audio_model = Model.get_radio_model()[3]['AUDIO']
        
        # Populate the grid
        self.__setup_ui(grid)
    
    #-------------------------------------------------
    # Populate the UI
    def __setup_ui(self, grid) :
        
        # Output sinks
        # Select HPSDR or Local
        src_label = QLabel('Sink')
        src_label.setStyleSheet("QLabel {color: rgb(196,196,196)}")
        grid.addWidget(src_label, 0, 0)
        self.__sink_combo = QComboBox()
        self.__sink_combo.addItems((HPSDR, LOCAL_AF, LOCAL_IQ))
        grid.addWidget(self.__sink_combo, 0, 1, 1, 3)
        self.__sink_combo.activated.connect(self.__sink_evnt)
        
        # Select the audio device
        dev_label = QLabel('Device')
        dev_label.setStyleSheet("QLabel {color: rgb(196,196,196)}")
        grid.addWidget(dev_label, 1, 0)
        self.__dev_combo = QComboBox()
        output_list = self.__con.cmd_exchange(M_ID.ENUM_OUTPUTS, [])
        devices = []
        if output_list == None:
            print("Failed to enumerate audio outputs!")
        else:
            devices.append('No selection')
            for device in output_list['outputs']:
                devices.append('%s@%s' % (device['api'], device['name']))
        self.__dev_combo.addItems(devices)
        grid.addWidget(self.__dev_combo, 1, 1, 1, 3)
        self.__dev_combo.activated.connect(self.__dev_evnt)
        
        # Assign channel
        # Create an exclusive button group
        self.btn_grp = QButtonGroup()
        self.btn_grp.setExclusive(True)
        btn_panel = QWidget()
        btn_box = QHBoxLayout()
        btn_panel.setLayout(btn_box)
        grid.addWidget(btn_panel, 2, 1, 1, 3)
        ch_label = QLabel('Ch')
        ch_label.setStyleSheet("QLabel {color: rgb(196,196,196)}")
        grid.addWidget(ch_label, 2, 0)
        
        self.__ch_left = QCheckBox('Left')
        self.btn_grp.addButton(self.__ch_left)
        self.btn_grp.setId(self.__ch_left, 0)
        self.__ch_left.setStyleSheet("QCheckBox {color: rgb(196,196,196)}")
        btn_box.addWidget(self.__ch_left)
        self.__ch_left.setVisible(False)
        
        self.__ch_right = QCheckBox('Right')
        self.btn_grp.addButton(self.__ch_right)
        self.btn_grp.setId(self.__ch_right, 1)
        self.__ch_right.setStyleSheet("QCheckBox {color: rgb(196,196,196)}")
        btn_box.addWidget(self.__ch_right)
        
        self.__ch_both = QCheckBox('Both')
        self.btn_grp.addButton(self.__ch_both)
        self.btn_grp.setId(self.__ch_both, 2)
        self.__ch_both.setStyleSheet("QCheckBox {color: rgb(196,196,196)}")
        btn_box.addWidget(self.__ch_both)
        
        self.__ch_none = QCheckBox('None')
        self.btn_grp.addButton(self.__ch_none)
        self.btn_grp.setId(self.__ch_none, 3)
        self.__ch_none.setStyleSheet("QCheckBox {color: rgb(196,196,196)}")
        btn_box.addWidget(self.__ch_none)
        
        self.btn_grp.buttonClicked.connect(self.__ch_evnt)
        
        # 'Apply' buttons
        applybtn = QPushButton('Apply')
        applybtn.setToolTip('Apply this audio configuration')
        applybtn.resize(applybtn.sizeHint())
        applybtn.setEnabled(True)
        grid.addWidget(applybtn, 3, 2)
        applybtn.clicked.connect(self.__apply_evnt)
        applybtn.setStyleSheet("QPushButton {background-color: rgb(59,59,59); color: rgb(196,196,196); font: bold 10px}")
        
        # 'Cancel' buttons
        cancelbtn = QPushButton('Cancel')
        cancelbtn.setToolTip('Cancel changes')
        cancelbtn.resize(cancelbtn.sizeHint())
        cancelbtn.setEnabled(True)
        grid.addWidget(cancelbtn, 3, 3)
        cancelbtn.clicked.connect(self.__cancel_evnt)
        cancelbtn.setStyleSheet("QPushButton {background-color: rgb(59,59,59); color: rgb(196,196,196); font: bold 10px}")
        
        # Set UI state
        self.__reset_state()
        
    #==============================================================================================
    # PUBLIC
    #==============================================================================================
    
    #-------------------------------------------------
    # Set context according to id
    def set_context(self, callback, x, y, direction):
        
        """
        Set Context
        
        Arguments:  
            callback    --  callback here with mode string
            x           --  x coord of main window
            y           --  y coord of main window
            direction   --  RX or TX
            
        """
        self.callback = callback
        self.direction = direction
        # Position at top right corner of invoking button

        self.move( x, y)
        # Set UI state
        self.__reset_state()
        
    #==============================================================================================
    # PRIVATE
    #==============================================================================================
    
    #-------------------------------------------------
    # UI Events    
    #-------------------------------------------------
    
    #-------------------------------------------------
    def __sink_evnt(self, idx) :
        """
        Sink update
        
        Arguments:  
            
            
        """
        self.__sink = self.__sink_combo.currentText()
        self.__update_state()
    
    #-------------------------------------------------
    def __dev_evnt(self, idx) :
        """
        Device update
        
        Arguments:  
            
            
        """
  
        self.__dev = self.__dev_combo.currentText()
        self.__update_state()
    
    #-------------------------------------------------
    def __ch_evnt(self, btn) :
        """
        Ch update, any checkbox
        
        Arguments:  
            
            
        """
        if self.btn_grp.id(btn) == 0:
            ch = LEFT
        elif self.btn_grp.id(btn) == 1:
            ch = RIGHT
        elif self.btn_grp.id(btn) == 2:
            ch = BOTH
        elif self.btn_grp.id(btn) == 3:
            ch = NONE
        self.__ch = ch
    
    #-------------------------------------------------
    def __apply_evnt(self, e) :
        """
        Any button click event
        
        Arguments:  
            
            
        """
        
        # Set the model
        radio = self.__get_model()
        radio['SINK'] = self.__sink
        radio['DEV'] = self.__dev
        radio['CH'] = self.__ch
        # Reset the server audio routes
        self.__con.set_audio_routes(True)
        # It's a popup
        self.close()
    
    #-------------------------------------------------
    def __cancel_evnt(self, e) :
        """
        Any button click event
        
        Arguments:  
            
            
        """
        
        self.__sink = None
        self.__dev = None
        self.__ch = None
        # It's a popup
        self.close()
    
    #-------------------------------------------------
    # Helpers
    #-------------------------------------------------
    
    #-------------------------------------------------
    def __get_model(self):
        if self.__id == RX_1:
            return self.__radio_1_audio_model
        elif self.__id == RX_2:
            return self.__radio_2_audio_model
        else:
            return self.__radio_3_audio_model
    
    #-------------------------------------------------
    def __get_other_models(self):
        if self.__id == RX_1:
            return self.__radio_2_audio_model, self.__radio_3_audio_model
        elif self.__id == RX_2:
            return self.__radio_1_audio_model, self.__radio_3_audio_model
        else:
            return self.__radio_1_audio_model, self.__radio_2_audio_model
        
    #-------------------------------------------------
    # Set UI to last model state
    def __reset_state(self):
        # Set holding values
        radio = self.__get_model()
        self.__sink = radio['SINK']
        self.__dev = radio['DEV']
        self.__ch = radio['CH']
        
        # Update UI to last known state
        # Sink
        self.__sink_combo.setCurrentIndex(self.__sink_combo.findText(self.__sink))
        # Device
        if radio['DEV'] != None:
            index = self.__dev_combo.findText(self.__dev)
        else:
            index = 0
        self.__dev_combo.setCurrentIndex(index)
        # Set widget context
        self.__manage_dev()
        self.__manage_ch()
    
    #-------------------------------------------------
    # Set UI to last local state
    def __update_state(self):
        self.__manage_dev()
        self.__manage_ch()
        
    #-------------------------------------------------
    # Set device visability
    def __manage_dev(self):
        if self.__sink == HPSDR:
            self.__dev_combo.setVisible(False)
        else:
            self.__dev_combo.setVisible(True)
    
    #-------------------------------------------------
    # Set channel visibility
    def __manage_ch(self):
        # NO, use id and other radios
        # Available channels
        # We have to check not only what the setting is for this receiver but what channels are available
        # by checking all other recevers.
        other_radio_1, other_radio_2 = self.__get_other_models()
        if self.__dev == NONE or self.__sink == HPSDR:
            # No device defined or not local so we can't set a channel
            self.__ch_left.setVisible(False)
            self.__ch_right.setVisible(False)
            self.__ch_both.setVisible(False)
            self.__ch_none.setVisible(False)
        else :
            # We have a device so check the saved channel
            # Set all visible first
            self.__ch_left.setVisible(True)
            self.__ch_right.setVisible(True)
            self.__ch_both.setVisible(True)
            self.__ch_none.setVisible(True)
            # Set the channel according to model
            if self.__ch == NONE:
                self.__ch_none.setCheckState(Qt.Checked)
            elif self.__ch == LEFT:
                self.__ch_left.setCheckState(Qt.Checked)                
            elif self.__ch == RIGHT:
                self.__ch_right.setCheckState(Qt.Checked)    
            elif self.__ch == BOTH:
                self.__ch_both.setCheckState(Qt.Checked)
            
            # Then turn off those which are allocated elsewhere
            if  other_radio_1['CH'] == LEFT or other_radio_2['CH'] == LEFT:
                if  self.__dev == other_radio_1['DEV'] or \
                    self.__dev == other_radio_2['DEV']:
                    self.__ch_left.setVisible(False)
                    self.__ch_both.setVisible(False)
                    if self.__ch == LEFT:
                        print("Audio problem, more than one channel on the same device allocated LEFT ch!")
            
            if  other_radio_1['CH'] == RIGHT or other_radio_2['CH'] == RIGHT:
                if  self.__dev == other_radio_1['DEV'] or \
                    self.__dev == other_radio_2['DEV']:
                    self.__ch_right.setVisible(False)
                    self.__ch_both.setVisible(False)
                    if self.__ch == RIGHT:
                        print("Audio problem, more than one channel on the same device allocated RIGHT ch!")
            
            if  other_radio_1['CH'] == BOTH or other_radio_2['CH'] == BOTH:
                if  self.__dev == other_radio_1['DEV'] or \
                    self.__dev == other_radio_2['DEV']:
                    self.__ch_left.setVisible(False)
                    self.__ch_right.setVisible(False)
                    self.__ch_both.setVisible(False)
                    if self.__ch == LEFT or self.__ch == RIGHT or self.__ch == BOTH:
                        print("Audio problem, more than one channel on the same device conflicts with BOTH ch!")
                        

File: src/cpp/FLConsole/FLConsole/src/Ui/Components/audio.cpp
/*
audio.cpp

Audio selection window implementation for the FLTK Console

Copyright (C) 2020 by G3UKB Bob Cowdery

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

The authors can be reached by email at:

	bob@bobcowdery.plus.com

*/

#include "../../Includes/includes.h"

//==============================================================================
// PUBLIC

//==============================================================================
// Audio Output Component Window
//==============================================================================

void apply_output_cb(Fl_Widget* w, void* user_data) {
	// Call down to our Audio instance
	((AudioOutput*)user_data)->handle_apply();
}

//----------------------------------------------------
// Constructor/Destructor
AudioOutput::AudioOutput(int radio, int w, int h) : Fl_Window(w, h) {

	// Our radio instance
	r = radio;

	// Get dependent objects from the cache
	r_i = (RadioInterface*)RSt::inst().get_obj("RADIO-IF");
	p = (Preferences*)RSt::inst().get_obj("PREFS");
	
	// Set up window
	char label[30];
	sprintf_s(label, "Audio Output (radio-%d", r);
	copy_label(label);
	resizable(this);
	color((Fl_Color)24);
	align(Fl_Align(65));
	
	// Add a group box
	Fl_Group *top_group = new Fl_Group(5, 5, w - 10, h - 10);
	top_group->box(FL_GTK_THIN_UP_BOX);
	top_group->color((Fl_Color)24);
	
	// Create a grid layout handler
	GridLayout *grid = new GridLayout(5, 5, w - 10, h - 10, 4, 5, 5);
	metrics m;
	// Add labels
	m = grid->get_cell_metrics(0, 0);
	Fl_Box* l_sink = new Fl_Box(m.x, m.y, m.w, m.h, "Sink");
	m = grid->get_cell_metrics(1, 0);
	Fl_Box* l_dev = new Fl_Box(m.x, m.y, m.w, m.h, "Dev");
	m = grid->get_cell_metrics(2, 0);
	Fl_Box* l_ch = new Fl_Box(m.x, m.y, m.w, m.h, "Ch");
	
	// Add choices
	m = grid->get_cell_metrics(0, 1, 1, 4);
	// Sink selection
	sink = new Fl_Choice(m.x, m.y, m.w, m.h);
	sink->add("HPSDR");
	sink->add("Local-AF");
	sink->add("Local-IQ");
	sink->value(0);
	m = grid->get_cell_metrics(1, 1, 1, 4);
	// Device selection
	device = new Fl_Choice(m.x, m.y, m.w, m.h);
	// Get the output enumeration
	DeviceEnumList* audio_outputs = c_server_enum_audio_outputs();
	// Populate the list
	char str[100];
	for (int i = 0; i < audio_outputs->entries; i++) {
		strcpy_s(str, 100, audio_outputs->devices[i].name);
		strcat_s(str, 100, ":");
		strcat_s(str, 100, audio_outputs->devices[i].host_api);
		device->add(str);
	}
	
	// Add channel selection
	m = grid->get_cell_metrics(2, 1, 1, 4);
	Fl_Group *ch_group = new Fl_Group(m.x, m.y, m.w, m.h);
	m = grid->get_cell_metrics(2, 1);
	left = new Fl_Radio_Light_Button(m.x, m.y, m.w, m.h, "Left");
	m = grid->get_cell_metrics(2, 2);
	right = new Fl_Radio_Light_Button(m.x, m.y, m.w, m.h, "Right");
	m = grid->get_cell_metrics(2, 3);
	both = new Fl_Radio_Light_Button(m.x, m.y, m.w, m.h, "Both");
	m = grid->get_cell_metrics(2, 4);
	none = new Fl_Radio_Light_Button(m.x, m.y, m.w, m.h, "None");
	ch_group->end();
	
	// Add apply button
	m = grid->get_cell_metrics(3, 4);
	apply = new Fl_Button(m.x, m.y, m.w, m.h, "Apply");
	apply->callback((Fl_Callback*)apply_output_cb, (void*)this);
	
	// Close up and display
	top_group->end();
	end();
	border(false);
	
	// Retrieve and reset audio path
	struct_audio_desc desc = p->get_audio_desc(r);
	if (desc.valid) {
		// Set the widget state
		set_widget_state(desc.loc_part, desc.api_part, desc.dev_part, desc.ch_part);
		// Set audio path
		r_i->set_audio_paths();
	}

	// Finally show window
	show();
}

//----------------------------------------------------
// Apply current state as a new audio path
void AudioOutput::handle_apply() {
	// Gather all the required information
	char sink_str[10];
	char dev_str[100];
	char *dev_part;
	char *api_part;
	char ch_str[10];

	// Set sink type
	if (strcmp(((Fl_Choice*)sink)->text(), "LOCAL-AF")) {
		strcpy_s(sink_str, 9, LOCAL_AF);
	} else if (strcmp(((Fl_Choice*)sink)->text(), "LOCAL-IQ")) {
		strcpy_s(sink_str, 9, LOCAL_IQ);
	} else if (strcmp(((Fl_Choice*)sink)->text(), "HPSDR")) {
		strcpy_s(sink_str, 9, HPSDR);
	}

	// Set dev type
	strcpy_s(dev_str, 100, ((Fl_Choice*)device)->text());
	// Split the dev part
	dev_part = strtok_s(dev_str, ":", &api_part);

	// Set channel type
	if (((Fl_Radio_Light_Button*)left)->value()) {
		strcpy_s(ch_str, 10, LEFT);
	}
	else if (((Fl_Radio_Light_Button*)right)->value()) {
		strcpy_s(ch_str, 10, RIGHT);
	}
	else {
		strcpy_s(ch_str, 10, BOTH);
	}

	// Save the new route
	save_route(r, sink_str, api_part, dev_part, ch_str);

	// Reset the audio path for all receivers 
	r_i->set_audio_paths();
}

//==============================================================================
// PRIVATE

//----------------------------------------------------
// Set a new audio path
void AudioOutput::set_widget_state(char* vsink, char* vapi, char* vdev, char* vch) {
	char str[100];

	// Set sink
	if (strcmp(vsink, LOCAL_AF) == 0) {
		((Fl_Choice*)sink)->value(((Fl_Choice*)sink)->find_index("Local-AF"));
	}
	else if (strcmp(vsink, LOCAL_IQ) == 0) {
		((Fl_Choice*)sink)->value(((Fl_Choice*)sink)->find_index("Local-IQ"));
	}
	else if (strcmp(vsink, HPSDR) == 0) {
		((Fl_Choice*)sink)->value(((Fl_Choice*)sink)->find_index("HPSDR"));
	}
	// Set device
	strcpy_s(str, 100, vdev);
	strcat_s(str, 100, ":");
	strcat_s(str, 100, vapi);
	((Fl_Choice*)device)->value(((Fl_Choice*)device)->find_index(str));
	// Set channel
	if (strcmp(vch, LEFT) == 0) {
		((Fl_Radio_Light_Button*)left)->set();
	}
	else if (strcmp(vch, RIGHT) == 0) {
		((Fl_Radio_Light_Button*)right)->set();
	}
	else if (strcmp(vch, BOTH) == 0) {
		((Fl_Radio_Light_Button*)both)->set();
	}
}

//----------------------------------------------------
// Save the audio route
void AudioOutput::save_route(int radio, char* sink, char* api, char* dev, char* ch) {
	// Save the current route
	char current_route[100];
	strcpy_s(current_route, 100, sink);
	strcat_s(current_route, 100, ":");
	strcat_s(current_route, 100, dev);
	strcat_s(current_route, 100, ":");
	strcat_s(current_route, 100, api);
	strcat_s(current_route, 100, ":");
	strcat_s(current_route, 100, ch);
	p->set_audio_path(radio, current_route);
}

//==============================================================================
// Audio Input Component Window
//==============================================================================

void apply_input_cb(Fl_Widget* w, void* user_data) {
	// Call down to our Audio instance
	((AudioInput*)user_data)->handle_apply();
}

//----------------------------------------------------
// Constructor/Destructor
AudioInput::AudioInput(int radio, int w, int h) : Fl_Window(w, h) {

	// Our radio instance
	r = radio;

	// Get dependent objects from the cache
	r_i = (RadioInterface*)RSt::inst().get_obj("RADIO-IF");
	p = (Preferences*)RSt::inst().get_obj("PREFS");

	// Set up window
	copy_label("Audio Input (transmitter)");
	resizable(this);
	color((Fl_Color)24);
	align(Fl_Align(65));

	// Add a group box
	Fl_Group *top_group = new Fl_Group(5, 5, w - 10, h - 10);
	top_group->box(FL_GTK_THIN_UP_BOX);
	top_group->color((Fl_Color)24);

	// Create a grid layout handler
	GridLayout *grid = new GridLayout(5, 5, w - 10, h - 10, 4, 5, 5);
	metrics m;
	// Add labels
	m = grid->get_cell_metrics(0, 0);
	Fl_Box* l_sink = new Fl_Box(m.x, m.y, m.w, m.h, "Source");
	m = grid->get_cell_metrics(1, 0);
	Fl_Box* l_dev = new Fl_Box(m.x, m.y, m.w, m.h, "Dev");

	// Add choices
	m = grid->get_cell_metrics(0, 1, 1, 4);
	// Source selection
	source = new Fl_Choice(m.x, m.y, m.w, m.h);
	source->add("HPSDR");
	source->add("Local");
	source->value(0);
	m = grid->get_cell_metrics(1, 1, 1, 4);
	// Device selection
	device = new Fl_Choice(m.x, m.y, m.w, m.h);
	// Get the output enumeration
	DeviceEnumList* audio_inputs = c_server_enum_audio_inputs();
	// Populate the list
	char str[100];
	for (int i = 0; i < audio_inputs->entries; i++) {
		strcpy_s(str, 100, audio_inputs->devices[i].name);
		strcat_s(str, 100, ":");
		strcat_s(str, 100, audio_inputs->devices[i].host_api);
		device->add(str);
	}

	// Add apply button
	m = grid->get_cell_metrics(3, 4);
	apply = new Fl_Button(m.x, m.y, m.w, m.h, "Apply");
	apply->callback((Fl_Callback*)apply_input_cb, (void*)this);

	// Close up and display
	top_group->end();
	end();
	border(false);

	// Retrieve and reset audio path
	struct_audio_desc desc = p->get_audio_desc(r);
	if (desc.valid) {
		// Set the widget state
		set_widget_state(desc.loc_part, desc.api_part, desc.dev_part);
		// Set audio path
		r_i->set_audio_paths();
	}

	// Finally show window
	show();
}

//----------------------------------------------------
// Apply current state as a new audio path
void AudioInput::handle_apply() {
	// Gather all the required information
	char source_str[10];
	char dev_str[100];
	char *dev_part;
	char *api_part;

	// Set source type
	if (strcmp(((Fl_Choice*)source)->text(), "LOCAL")) {
		strcpy_s(source_str, 9, LOCAL_AF);
	}
	else if (strcmp(((Fl_Choice*)source)->text(), "HPSDR")) {
		strcpy_s(source_str, 9, HPSDR);
	}

	// Set dev type
	strcpy_s(dev_str, 100, ((Fl_Choice*)device)->text());
	// Split the dev part
	dev_part = strtok_s(dev_str, ":", &api_part);

	// Save the new route
	save_route(r, source_str, api_part, dev_part);

	// Reset the audio path for all receivers 
	r_i->set_audio_paths();
}

//==============================================================================
// PRIVATE

//----------------------------------------------------
// Set a new audio path
void AudioInput::set_widget_state(char* vsource, char* vapi, char* vdev) {
	char str[100];

	// Set source
	if (strcmp(vsource, LOCAL) == 0) {
		((Fl_Choice*)source)->value(((Fl_Choice*)source)->find_index("Local-IQ"));
	}
	else if (strcmp(vsource, HPSDR) == 0) {
		((Fl_Choice*)source)->value(((Fl_Choice*)source)->find_index("HPSDR"));
	}
	// Set device
	strcpy_s(str, 100, vdev);
	strcat_s(str, 100, ":");
	strcat_s(str, 100, vapi);
	((Fl_Choice*)device)->value(((Fl_Choice*)device)->find_index(str));
}

//----------------------------------------------------
// Save the audio route
void AudioInput::save_route(int radio, char* source, char* api, char* dev) {
	// Save the current route
	char current_route[100];
	strcpy_s(current_route, 100, source);
	strcat_s(current_route, 100, ":");
	strcat_s(current_route, 100, dev);
	strcat_s(current_route, 100, ":");
	strcat_s(current_route, 100, api);
	strcat_s(current_route, 100, ":BOTH");
	p->set_audio_path(radio, current_route);
}
